clients/drdisasm/drdisasm.cpp

```diff
@@ -0,0 +1,158 @@
+/* **********************************************************
+ * Copyright (c) 2020 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#define DR_FAST_IR 1
+#include "dr_api.h"
+#include "droption.h"
+#include <iostream>
+#include <sstream>
+
+namespace {
+
+// XXX i#1684: We want cross-arch decoding support so a single build can decode
+// AArchXX and x86.  For now, a separate build is needed.
+#ifdef X86_64
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "x64",
+                                "Decodes using the specified mode: 'x64' or 'x86'.",
+                                "Decodes using the specified mode: 'x64' or 'x86'.");
+#elif defined(ARM)
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "arm",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.");
+#endif
+droption_t<bool> op_show_bytes(DROPTION_SCOPE_FRONTEND, "show_bytes", true,
+                               "Display the instruction encoding bytes.",
+                               "Display the instruction encoding bytes.");
+
+static bool
+parse_bytes(std::string token, std::vector<byte> &bytes)
+{
+    // Assume everything is hex.
+    // Assume consecutive values should be split and endianized.
```

Comment:
But there is a do while loop to handle consecutive hex values at line 69. Why are you making this assumption? I think for cases where the user enter "0x140x320x110x99" right?

Reply:
The comment could be clearer: it means that we assume that a larger-than-single-byte-value like a whole word or dword (like 0xdeadbeef) should be split as though it were a machine word and thus for x86 is little-endian, rather than just strung together left-to-right.  I.e., treat "9066" as "66 90" on x86.  This is definitely what we want to do for aarchxx -- for x86 it is maybe less clear but I still think it's better this way.  Agreed?

Reply:
To elaborate for aarchxx: for Thumb the encoding is typically written as two separate 2-byte little-endian words, while for Arm and AArch64 it is written as one 4-byte little-endian word.  E.g., from the test template:
```
#if defined(AARCH64)
 f94017a0   ldr    +0x28(%x29)[8byte] -> %x0
 a9be7bfd   stp    %x29 %x30 %sp $0xffffffffffffffe0 -> -0x20(%sp)[16byte] %sp
#elif defined(ARM) && !defined(thumb)
 f2436813   vtst.8 %d3 %d3 -> %d22
 b2db42c8   sbcs.lt %r11 $0x8000000c -> %r4
#elif defined(ARM) && defined(thumb)
 6813       ldr    (%r2)[4byte] -> %r3
 f243 42c8  movw   $0x000034c8 -> %r2
 b2db       uxtb   %r3[1byte] -> %r3
```

For x86, I figure you might want to decode some memory that you obtained from `od` or `gdb` in units larger than bytes, and you would also want it treated as little-endian.

Reply:
Ah, I see what you mean now. Yes, I agree.

Reply:
OK, I will tweak the comment to make this more clear.